File: swift-bridge/Sources/SpriteKitBridge/Region.swift
import SpriteKit

@_cdecl("sk_region_infinite")
public func sk_region_infinite() -> UnsafeMutableRawPointer? {
    skRetain(SKRegion.infinite())
}

@_cdecl("sk_region_new_with_radius")
public func sk_region_new_with_radius(_ radius: Float) -> UnsafeMutableRawPointer? {
    skRetain(SKRegion(radius: radius))
}

@_cdecl("sk_region_new_with_size")
public func sk_region_new_with_size(_ width: Double, _ height: Double) -> UnsafeMutableRawPointer? {
    skRetain(SKRegion(size: CGSize(width: width, height: height)))
}

@_cdecl("sk_region_inverse")
public func sk_region_inverse(_ regionHandle: UnsafeMutableRawPointer?) -> UnsafeMutableRawPointer? {
    guard let region: SKRegion = skBorrow(regionHandle) else { return nil }
    return skRetain(region.inverse())
}

@_cdecl("sk_region_union")
public func sk_region_union(_ regionHandle: UnsafeMutableRawPointer?, _ otherHandle: UnsafeMutableRawPointer?) -> UnsafeMutableRawPointer? {
    guard let region: SKRegion = skBorrow(regionHandle),
          let other: SKRegion = skBorrow(otherHandle)
    else { return nil }
    return skRetain(region.byUnion(with: other))
}

@_cdecl("sk_region_difference")
public func sk_region_difference(_ regionHandle: UnsafeMutableRawPointer?, _ otherHandle: UnsafeMutableRawPointer?) -> UnsafeMutableRawPointer? {
    guard let region: SKRegion = skBorrow(regionHandle),
          let other: SKRegion = skBorrow(otherHandle)
    else { return nil }
    return skRetain(region.byDifference(from: other))
}

@_cdecl("sk_region_intersection")
public func sk_region_intersection(_ regionHandle: UnsafeMutableRawPointer?, _ otherHandle: UnsafeMutableRawPointer?) -> UnsafeMutableRawPointer? {
    guard let region: SKRegion = skBorrow(regionHandle),
          let other: SKRegion = skBorrow(otherHandle)
    else { return nil }
    return skRetain(region.byIntersection(with: other))
}

@_cdecl("sk_region_contains_point")
public func sk_region_contains_point(_ regionHandle: UnsafeMutableRawPointer?, _ x: Double, _ y: Double) -> Bool {
    guard let region: SKRegion = skBorrow(regionHandle) else { return false }
    return region.contains(CGPoint(x: x, y: y))
}

File: src/region.rs
use apple_cf::cg::{CGPoint, CGSize};

use crate::ffi;
use crate::private::handle_type;

handle_type!(Region);

impl Region {
    /// Wraps `SKRegion`.
    #[must_use]
    pub fn infinite() -> Option<Self> {
        unsafe { Self::from_raw(ffi::sk_region_infinite()) }
    }

    /// Wraps `SKRegion`.
    #[must_use]
    pub fn with_radius(radius: f32) -> Option<Self> {
        unsafe { Self::from_raw(ffi::sk_region_new_with_radius(radius)) }
    }

    /// Wraps `SKRegion`.
    #[must_use]
    pub fn with_size(size: CGSize) -> Option<Self> {
        unsafe { Self::from_raw(ffi::sk_region_new_with_size(size.width, size.height)) }
    }

    /// Wraps `SKRegion`.
    #[must_use]
    pub fn inverse(&self) -> Option<Self> {
        unsafe { Self::from_raw(ffi::sk_region_inverse(self.ptr)) }
    }

    /// Wraps `SKRegion`.
    #[must_use]
    pub fn union(&self, other: &Self) -> Option<Self> {
        unsafe { Self::from_raw(ffi::sk_region_union(self.ptr, other.ptr)) }
    }

    /// Wraps `SKRegion`.
    #[must_use]
    pub fn difference(&self, other: &Self) -> Option<Self> {
        unsafe { Self::from_raw(ffi::sk_region_difference(self.ptr, other.ptr)) }
    }

    /// Wraps `SKRegion`.
    #[must_use]
    pub fn intersection(&self, other: &Self) -> Option<Self> {
        unsafe { Self::from_raw(ffi::sk_region_intersection(self.ptr, other.ptr)) }
    }

    /// Wraps `SKRegion`.
    #[must_use]
    pub fn contains_point(&self, point: CGPoint) -> bool {
        unsafe { ffi::sk_region_contains_point(self.ptr, point.x, point.y) }
    }
}
